toolkits/jira/arcade_jira/utils.py

```diff
@@ -1123,22 +1190,84 @@ def extract_id(field: Any) -> dict[str, str] | None:
     return {"id": field["id"]} if isinstance(field, dict) else None
 
 
-def build_issue_url(cloud_name: str | None, issue_id: str, issue_key: str) -> str | None:
-    if not cloud_name:
-        return None
+async def resolve_cloud_id(context: ToolContext, cloud_id: str | None) -> str:
+    if isinstance(cloud_id, str) and cloud_id != "":
+        return cloud_id
 
-    return f"https://{cloud_name}.atlassian.net/jira/software/projects/{issue_id}/list?selectedIssue={issue_key}"
+    from arcade_jira.tools.cloud import get_available_atlassian_clouds  # Avoid circular import
 
+    response = await get_available_atlassian_clouds(context)
+    clouds = response["clouds_available"]
 
-def build_project_url(cloud_name: str | None, project_key: str) -> str | None:
-    if not cloud_name:
-        return None
+    if len(clouds) == 0:
+        message = "No Atlassian Cloud is available. Please authorize an Atlassian Cloud."
+        raise ToolExecutionError(
+            message=message,
+            developer_message=message,
+        )
 
-    return f"https://{cloud_name}.atlassian.net/jira/software/projects/{project_key}/summary"
+    if len(clouds) > 1:
+        message = (
+            "Multiple Atlassian Clouds are available. One Cloud ID has to be selected and provided "
+            "in the tool call using the `atlassian_cloud_id` argument."
+        )
+        raise RetryableToolError(
```

Comment:
This is great. I was in the chat playground and said "list all of my projects across my workspaces" without providing any context about cloud_ids, and the model ended up taking the context from the error and calling `list_projects` once for each cloud_id (3 times).